### PR TITLE
Sequence model types refactor

### PR DESCRIPTION
**Pull request by MisterVants**
This refactor aims to improve the model graph of the SDK, simplify the handling + scaling of codable types, and provide a proposal of distinction for public and internal models.

It should help scaling tests and making stubbing data simpler and more straightforward on the long-term.

**Public API proposal for model types:**
- All public, session-related types, should fall under the public `Session` namespace umbrella.
- Bridging `init` methods can be provided to interchange between public and internal types when necessary.

@llbartekll these are big changes, I believe it will help scaling the code soon. Tell me if you have any insights about.

### CODE REVIEW

**Comment by llbartekll** [comment]
> All public, session-related types, should fall under the public Session namespace umbrella.

would it make sense now to move `SessionType.Reason` to `Session` namespace and make `SessionType` internal? it could be also applied to `PairingType`

**Comment by MisterVants** [comment]
Yesssss, exactly, you got the point!! That's one of the next things I'm going for

**Review by llbartekll — CHANGES_REQUESTED** [review]
1) I like public Session namespace but do not understand why you extracted so many internal types(Client Communication) from `SessonType` namespace.
2) Leaning the engines looks very good
3) I think we cannot have so many `CryptoKit` dependencies. Those modules should depend on abstraction. IMHO previous approach was better.
4) Only two types of WC responses are handled but I believe it is postponed for later
5) your approach for WCrelay looks good

File: Sources/WalletConnect/Engine/SessionEngine.swift [hunk 1]
```diff
@@ -566,4 +515,59 @@ final class SessionEngine {
...
+    private func handleReponse(_ response: WCResponse) {
+        switch response.requestParams {
+        case .pairingPayload(let payloadParams):
+            let proposeParams = payloadParams.request.params
+            handleProposeResponse(topic: response.topic, proposeParams: proposeParams, result: response.result)
+        case .sessionApprove(let approveParams):
+            handleApproveResponse(topic: response.topic, result: response.result)
+        default:
+            break
+        }
```
[the remark this answers is not in the document]
Comment (MisterVants) [reply Sources/WalletConnect/Engine/SessionEngine.swift:528]: Yes, they can just be added as new cases

File: Sources/WalletConnect/Engine/SessionEngine.swift [hunk 2]
```diff
@@ -566,4 +515,59 @@ final class SessionEngine {
...
+                    methods: proposal.permissions.jsonrpc.methods))
+            onApprovalAcknowledgement?(sessionSuccess)
+        case .failure:
+            wcSubscriber.removeSubscription(topic: topic)
+            wcSubscriber.removeSubscription(topic: settledTopic)
+            sequencesStore.delete(topic: topic)
+            sequencesStore.delete(topic: settledTopic)
+            crypto.deleteAgreementKeys(for: topic)
+            crypto.deleteAgreementKeys(for: settledTopic)
+            crypto.deletePrivateKey(for: pendingSession.publicKey)
```
[the remark this answers is not in the document]
Comment (MisterVants) [reply Sources/WalletConnect/Engine/SessionEngine.swift:570]: I think it's the only way to let the client know that the approval failed, we should add it later

File: Sources/WalletConnect/Crypto/Crypto.swift [hunk 3]
```diff
@@ -22,40 +45,32 @@ class Crypto: CryptoStorageProtocol {
...
-    func set(agreementKeys: Crypto.X25519.AgreementKeys, topic: String) {
-        let agreement = agreementKeys.sharedSecret + agreementKeys.publicKey
-        do {
-            try keychain.add(agreement, forKey: topic)
-        } catch {
-            print("Error adding agreement keys: \(error)")
+
+    func getPrivateKey(for publicKey: Curve25519.KeyAgreement.PublicKey) throws -> Curve25519.KeyAgreement.PrivateKey? {
+        guard let privateKeyData = try? keychain.read(key: publicKey.rawRepresentation.toHexString()) as Data else {
+            return nil
```
[the remark this answers is not in the document]
Comment (MisterVants) [reply Sources/WalletConnect/Crypto/Crypto.swift:58]: Agreed, I'll check a solution

File: Sources/WalletConnect/Types/ClientSynchJSONRPC/ClientSynchJSONRPC_Params.swift [hunk 4]
```diff
@@ -3,7 +3,7 @@ import Foundation
 
 extension WCRequest {
     enum Params: Codable, Equatable {
-        case pairingApprove(PairingType.ApproveParams)
+        case pairingApprove(PairingApproval)
         case pairingReject(PairingType.RejectParams)
         case pairingUpdate(PairingType.UpdateParams)
         case pairingUpgrade(PairingType.UpgradeParams)
```
[the remark this answers is not in the document]
Comment (MisterVants) [reply Sources/WalletConnect/Types/ClientSynchJSONRPC/ClientSynchJSONRPC_Params.swift:9]: I think I've written this before deciding what to do with the `Session` sub-types. I need to revisit it.

**Pull request merged**